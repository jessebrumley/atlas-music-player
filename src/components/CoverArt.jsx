import placeholder from "../assets/placeholder.svg";

export default function CoverArt() {
  return (
    <img
      src={placeholder}
      alt="Cover art"
      className="w-full aspect-square rounded-xl object-cover"
    />
  );
}
