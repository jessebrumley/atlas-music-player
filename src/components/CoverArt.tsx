/**
 * CoverArt.tsx
 *
 * Renders the cover image for a song.
 * If no `cover` prop is provided or the image fails to load,
 * a default placeholder image is used instead.
 *
 * When the user hovers over the image, lyrics are fetched from the API
 * using the song ID and displayed as an overlay.
 */
import { useEffect, useState } from "react";
import placeholder from "../assets/placeholder.svg";

interface CoverArtProps {
  cover?: string;
  id?: string;
}

export default function CoverArt({ cover, id }: CoverArtProps) {
  const [lyrics, setLyrics] = useState<string | null>(null);
  const [showLyrics, setShowLyrics] = useState(false);

  useEffect(() => {
    const fetchLyrics = async () => {
      if (!id) return;
      try {
        const res = await fetch(`/api/v1/lyrics/${id}`);
        if (!res.ok) throw new Error("Lyrics not found");
        const data = await res.json();
        setLyrics(data.lyrics);
      } catch (err) {
        setLyrics(null);
      }
    };

    setLyrics(null);
    fetchLyrics();
  }, [id]);

  return (
    <div
      className="relative"
      onMouseEnter={() => setShowLyrics(true)}
      onMouseLeave={() => setShowLyrics(false)}
    >
      <img
        src={cover || placeholder}
        alt="Cover Art"
        className="rounded-lg drop-shadow-md"
        onError={(e) => {
          e.currentTarget.onerror = null;
          e.currentTarget.src = placeholder;
        }}
      />
      {showLyrics && lyrics && (
        <div className="absolute inset-0 overflow-auto whitespace-pre-wrap rounded-lg bg-background bg-opacity-90 p-4 text-sm text-text z-10">
          {lyrics}
        </div>
      )}
    </div>
  );
}
