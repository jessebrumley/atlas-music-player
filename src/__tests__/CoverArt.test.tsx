import "@testing-library/jest-dom/vitest";
import { describe, it, expect, beforeAll, afterEach, afterAll } from "vitest";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import CoverArt from "../components/CoverArt";

const mockCover = "https://example.com/cover.jpg";
const mockLyrics = "These are test lyrics.";
const mockId = "test-song-id";

describe("CoverArt", () => {
  it("renders with provided cover image", () => {
    render(<CoverArt cover={mockCover} />);
    const img = screen.getByAltText(/cover art/i);
    expect(img).toHaveAttribute("src", mockCover);
  });

  it("renders placeholder if no cover provided", () => {
    render(<CoverArt />);
    const img = screen.getByAltText(/cover art/i);
    expect(img.src).toContain("data:image/svg+xml");
  });

  it("displays lyrics on hover after fetching", async () => {
    render(<CoverArt id={mockId} />);
    const container = screen.getByRole("img", { name: /cover art/i }).parentElement!;
    fireEvent.mouseEnter(container);

    await waitFor(() => {
      expect(screen.getByText(mockLyrics)).toBeInTheDocument();
    });

    fireEvent.mouseLeave(container);
    await waitFor(() => {
      expect(screen.queryByText(mockLyrics)).not.toBeInTheDocument();
    });
  });

  it("matches snapshot", () => {
    const { container } = render(<CoverArt cover={mockCover} />);
    expect(container).toMatchSnapshot();
  });
});
