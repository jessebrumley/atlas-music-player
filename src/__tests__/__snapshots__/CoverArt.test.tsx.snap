// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CoverArt > matches snapshot 1`] = `
<div>
  <div
    class="relative"
  >
    <img
      alt="Cover Art"
      class="rounded-lg drop-shadow-md"
      src="https://example.com/cover.jpg"
    />
  </div>
</div>
`;
